File: infer/tests/codetoanalyze/java/pulse/taint/Interprocedural.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

class Interprocedural {

  Object f;

  static Object sGlobal;

  static class Obj {
    Object f;
  }

  public static Object id(Object param) {
    return param;
  }

  /** source tests */
  public static Object returnSourceDirect() {
    return InferTaint.inferSecretSource();
  }

  public static Object returnSourceIndirect() {
    return returnSourceDirect();
  }

  public static void returnSourceDirectBad() {
    InferTaint.inferSensitiveSink(returnSourceDirect());
  }

  public static void returnSourceDirectViaVarBad() {
    Object source = returnSourceDirect();
    InferTaint.inferSensitiveSink(source);
  }

  public static void returnSourceIndirectBad() {
    InferTaint.inferSensitiveSink(returnSourceIndirect());
  }

  public static Obj returnSourceViaField() {
    Obj o = new Obj();
    o.f = InferTaint.inferSecretSource();
    return o;
  }

  public static void returnSourceViaFieldBad() {
    InferTaint.inferSensitiveSink(returnSourceViaField().f);
  }

  public static void returnSourceViaParameter1(Obj o) {
    o.f = InferTaint.inferSecretSource();
  }

  public static void returnSourceViaParameter1Bad(Obj o) {
    returnSourceViaParameter1(o);
    InferTaint.inferSensitiveSink(o.f);
  }

  public static void returnSourceViaParameter2(Obj o1, Obj o2) {
    o2.f = o1.f;
  }

  public static void returnSourceViaParameter2Bad(Obj o1, Obj o2) {
    o1.f = InferTaint.inferSecretSource();
    returnSourceViaParameter2(o1, o2);
    InferTaint.inferSensitiveSink(o2.f);
  }

  public static void returnSourceViaParameterOk(Obj o1, Obj o2) {
    o1.f = InferTaint.inferSecretSource();
    returnSourceViaParameter2(o2, o1);
    InferTaint.inferSensitiveSink(o2.f);
  }

  public static void returnSourceViaGlobal() {
    sGlobal = InferTaint.inferSecretSource();
  }

  public void returnSourceViaGlobalBad() {
    returnSourceViaGlobal();
    InferTaint.inferSensitiveSink(sGlobal);
  }

  public void returnSourceViaGlobalOk() {
    returnSourceViaGlobal();
    sGlobal = null;
    InferTaint.inferSensitiveSink(sGlobal);
  }

  /** sink tests */
  public static void callSinkParam1(Object param1, Object param2) {
    InferTaint.inferSensitiveSink(param1);
  }

  public static void callSinkParam1Bad() {
    callSinkParam1(InferTaint.inferSecretSource(), null);
  }

  public static void callSinkParam1Ok() {
    callSinkParam1(null, InferTaint.inferSecretSource());
  }

  public static void callSinkParam2(Object param1, Object param2) {
    InferTaint.inferSensitiveSink(param2);
  }

  public static void callSinkParam2Bad() {
    callSinkParam2(null, InferTaint.inferSecretSource());
  }

  public static void callSinkParam2Ok() {
    callSinkParam2(InferTaint.inferSecretSource(), null);
  }

  public void callSinkOnFieldDirect() {
    InferTaint.inferSensitiveSink(this.f);
  }

  public void callSinkOnFieldDirectBad() {
    this.f = InferTaint.inferSecretSource();
    callSinkOnFieldDirect();
  }

  public static void callSinkOnFieldIndirect(Obj param) {
    InferTaint.inferSensitiveSink(param.f);
  }

  public static void callSinkOnFieldIndirectBad() {
    Obj obj = new Obj();
    obj.f = InferTaint.inferSecretSource();
    callSinkOnFieldIndirect(obj);
  }

  public Object getF() {
    return f;
  }

  void callSinkOnLocal() {
    Object local = this.getF();
    InferTaint.inferSensitiveSink(local);
  }

  void callSinkOnLocalBad() {
    this.f = InferTaint.inferSecretSource();
    callSinkOnLocal();
  }

  public static void callSinkOnGlobal() {
    InferTaint.inferSensitiveSink(sGlobal);
  }

  public static void callSinkOnGlobalBad() {
    sGlobal = InferTaint.inferSecretSource();
    callSinkOnGlobal();
  }

  public static void callSinkOnGlobalOk() {
    sGlobal = InferTaint.inferSecretSource();
    sGlobal = null;
    callSinkOnGlobal();
  }

  public static void setGlobal(Object o) {
    sGlobal = o;
  }

  public static void setGlobalThenCallSinkBad() {
    setGlobal(InferTaint.inferSecretSource());
    callSinkOnGlobal();
  }

  public static Object getGlobal() {
    return sGlobal;
  }

  public static void getGlobalThenCallSink() {
    Object local = getGlobal();
    InferTaint.inferSensitiveSink(sGlobal);
  }

  public static void getGlobalThenCallSinkBad() {
    sGlobal = InferTaint.inferSecretSource();
    getGlobalThenCallSink();
  }

  // this should report two alarms, not three
  public void callSinkNoTripleReportBad() {
    Object source = InferTaint.inferSecretSource();
    callSinkParam1(source, null);
    callSinkParam2(null, source);
  }

  /** passthrough tests */
  public static void singlePassthroughBad() {
    Object source = InferTaint.inferSecretSource();
    Object launderedSource = id(source);
    InferTaint.inferSensitiveSink(launderedSource);
  }

  public static void doublePassthroughBad() {
    Object source = InferTaint.inferSecretSource();
    Object launderedSource1 = id(source);
    Object launderedSource2 = id(launderedSource1);
    InferTaint.inferSensitiveSink(launderedSource2);
  }

  /** false positives: an ideal analysis would not report these, but we will */
  public static Object returnSourceConditional(boolean b) {
    if (b) return InferTaint.inferSecretSource();
    return null;
  }

  public static void trackParamsOk() {
    InferTaint.inferSensitiveSink(returnSourceConditional(false));
  }

  public static void reassignInCallee(Obj o) {
    o.f = null;
  }

  public static void reassignInCallee() {
    Obj o = new Obj();
    o.f = InferTaint.inferSecretSource();
    reassignInCallee(o);
    InferTaint.inferSensitiveSink(o.f);
  }

  static Object relevantPassthrough(Object param) {
    return param;
  }

  static Object irrelevantPassthrough(Object param) {
    return param;
  }

  // the following tests should show only "relevantPassthrough" in their traces
  public static Object irrelevantPassthroughsIntraprocedural(Object param) {
    Object irrelevant = irrelevantPassthrough(param);
    Object source = InferTaint.inferSecretSource();
    Object relevant = relevantPassthrough(source);
    InferTaint.inferSensitiveSink(relevant);
    return irrelevantPassthrough(relevant);
  }

  public static Object returnSourceIrrelevantPassthrough(Object param) {
    Object irrelevant = irrelevantPassthrough(param);
    Object source = InferTaint.inferSecretSource();
    return relevantPassthrough(source);
  }

  public static Object irrelevantPassthroughsSourceInterprocedural(Object o) {
    Object irrelevant = irrelevantPassthrough(o);
    Object source = returnSourceIrrelevantPassthrough(irrelevant);
    Object relevant = relevantPassthrough(source);
    InferTaint.inferSensitiveSink(relevant);
    return irrelevantPassthrough(source);
  }

  public static Object callSinkIrrelevantPassthrough(Object param) {
    Object relevant = relevantPassthrough(param);
    InferTaint.inferSensitiveSink(relevant);
    Object irrelevant = irrelevantPassthrough(param);
    return irrelevant;
  }

  public static Object irrelevantPassthroughsSinkInterprocedural(Object o) {
    Object source = InferTaint.inferSecretSource();
    Object relevant = relevantPassthrough(source);
    callSinkIrrelevantPassthrough(relevant);
    return irrelevantPassthrough(relevant);
  }

  public static Object irrelevantPassthroughsSourceAndSinkInterprocedural(Object o) {
    Object irrelevant = irrelevantPassthrough(o);
    Object source = returnSourceIrrelevantPassthrough(irrelevant);
    Object relevant = relevantPassthrough(source);
    callSinkIrrelevantPassthrough(relevant);
    return irrelevantPassthrough(relevant);
  }

  public static void callSinkVariadic(Object... params) {
    InferTaint.inferSensitiveSink(params);
  }

  public static void callSinkVariadicBad() {
    callSinkVariadic(null, null, InferTaint.inferSecretSource());
  }

  void diverge() {
    for (; ; ) ;
  }

  public void divergenceInCallee() {
    Object source = InferTaint.inferSecretSource();
    diverge();
    InferTaint.inferSensitiveSink(source);
  }

  public static void callSinkThenDiverge(Object param) {
    InferTaint.inferSensitiveSink(param);
    for (; ; ) ;
  }

  public void FN_callSinkThenDivergeBad() {
    callSinkThenDiverge(InferTaint.inferSecretSource());
  }

  public void callSinkOnParam(Object o) {
    InferTaint.inferSensitiveSink(o);
  }

  public void callSinkIndirectOnParam(Object o) {
    callSinkOnParam(o);
  }

  Obj propagate(Object param) {
    Obj o = new Obj();
    o.f = param;
    return o;
  }

  static Obj id2(Obj o) {
    return o;
  }

  void callSinkA(Obj o) {
    callSink1(o);
  }

  void callSinkB(Obj o) {
    callSink2(o);
  }

  void callSinkC(Obj o) {
    callSink3(o);
  }

  void callSinkD(Obj o) {
    callSink4(o);
  }

  void callSink1(Obj o) {
    InferTaint.inferSensitiveSink(id(o));
  }

  void callSink2(Obj o) {
    InferTaint.inferSensitiveSink(id2(o).f);
  }

  void callSink3(Obj o) {
    InferTaint.inferSensitiveSink(id(o.f));
  }

  void callSink4(Obj o) {
    InferTaint.inferSensitiveSink(o.f);
  }

  public void callDeepSinkIndirectBad() {
    Object source = InferTaint.inferSecretSource();
    callSinkIndirectOnParam(source);
  }

  // the sink is on source but source.f is tainted, not source
  public void FN_callDeepSink1Bad() {
    Obj source = propagate(InferTaint.inferSecretSource());
    callSinkA(source);
  }

  public void callDeepSink2Bad() {
    Obj source = propagate(InferTaint.inferSecretSource());
    callSinkB(source);
  }

  void callShallowSinkBad(Obj o) {
    o.f = InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(id2(o).f);
  }

  public void callDeepSink3Bad() {
    Obj source = propagate(InferTaint.inferSecretSource());
    callSinkC(source);
  }

  public void callDeepSink4Bad() {
    Obj source = propagate(InferTaint.inferSecretSource());
    callSinkD(source);
  }

  public static void swapParams(Object o1, Object o2) {
    o1 = o2;
  }

  public static void assignSourceToParam(Object o) {
    o = InferTaint.inferSecretSource();
  }

  public static void swapParamsOk() {
    Object notASource = null;
    Object source = InferTaint.inferSecretSource();
    swapParams(notASource, source);
    InferTaint.inferSensitiveSink(notASource);
  }

  public static void assignSourceToParamOk() {
    Object o = null;
    assignSourceToParam(o);
    InferTaint.inferSensitiveSink(o);
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/Interprocedural.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

class Interprocedural {

  var f: Any? = null

  class Obj {
    var f: Any? = null
  }

  fun returnSourceViaGlobalBad() {
    returnSourceViaGlobal()
    InferTaint.inferSensitiveSink(global)
  }

  fun returnSourceViaGlobalOk() {
    returnSourceViaGlobal()
    global = null
    InferTaint.inferSensitiveSink(global)
  }

  fun callSinkOnFieldDirect() {
    InferTaint.inferSensitiveSink(f)
  }

  fun callSinkOnFieldDirectBad() {
    f = InferTaint.inferSecretSource()
    callSinkOnFieldDirect()
  }

  fun callSinkOnLocal() {
    val local = f
    InferTaint.inferSensitiveSink(local)
  }

  fun callSinkOnLocalBad() {
    f = InferTaint.inferSecretSource()
    callSinkOnLocal()
  }

  // this should report two alarms, not three
  fun callSinkNoTripleReportBad() {
    val source = InferTaint.inferSecretSource()
    callSinkParam1(source, null)
    callSinkParam2(null, source)
  }

  fun diverge() {
    while (true) {}
  }

  fun divergenceInCallee() {
    val source = InferTaint.inferSecretSource()
    diverge()
    InferTaint.inferSensitiveSink(source)
  }

  fun FN_callSinkThenDivergeBad() {
    callSinkThenDiverge(InferTaint.inferSecretSource())
  }

  fun callSinkOnParam(o: Any?) {
    InferTaint.inferSensitiveSink(o)
  }

  fun callSinkIndirectOnParam(o: Any?) {
    callSinkOnParam(o)
  }

  fun propagate(param: Any?): Obj {
    val o = Obj()
    o.f = param
    return o
  }

  fun callSinkA(o: Obj?) {
    callSink1(o)
  }

  fun callSinkB(o: Obj) {
    callSink2(o)
  }

  fun callSinkC(o: Obj) {
    callSink3(o)
  }

  fun callSinkD(o: Obj) {
    callSink4(o)
  }

  fun callSink1(o: Obj?) {
    InferTaint.inferSensitiveSink(id(o))
  }

  fun callSink2(o: Obj) {
    InferTaint.inferSensitiveSink(id2(o).f)
  }

  fun callSink3(o: Obj) {
    InferTaint.inferSensitiveSink(id(o.f))
  }

  fun callSink4(o: Obj) {
    InferTaint.inferSensitiveSink(o.f)
  }

  fun callDeepSinkIndirectBad() {
    val source = InferTaint.inferSecretSource()
    callSinkIndirectOnParam(source)
  }

  // the sink is on source but source.f is tainted, not source
  fun FN_callDeepSink1Bad() {
    val source = propagate(InferTaint.inferSecretSource())
    callSinkA(source)
  }

  fun callDeepSink2Bad() {
    val source = propagate(InferTaint.inferSecretSource())
    callSinkB(source)
  }

  fun callShallowSinkBad(o: Obj) {
    o.f = InferTaint.inferSecretSource()
    InferTaint.inferSensitiveSink(id2(o).f)
  }

  fun callDeepSink3Bad() {
    val source = propagate(InferTaint.inferSecretSource())
    callSinkC(source)
  }

  fun callDeepSink4Bad() {
    val source = propagate(InferTaint.inferSecretSource())
    callSinkD(source)
  }

  companion object {
    var global: Any? = null

    fun id(param: Any?): Any? {
      return param
    }

    /** source tests */
    fun returnSourceDirect(): Any {
      return InferTaint.inferSecretSource()
    }

    fun returnSourceIndirect(): Any {
      return returnSourceDirect()
    }

    fun returnSourceDirectBad() {
      InferTaint.inferSensitiveSink(returnSourceDirect())
    }

    fun returnSourceDirectViaVarBad() {
      val source = returnSourceDirect()
      InferTaint.inferSensitiveSink(source)
    }

    fun returnSourceIndirectBad() {
      InferTaint.inferSensitiveSink(returnSourceIndirect())
    }

    fun returnSourceViaField(): Obj {
      val o = Obj()
      o.f = InferTaint.inferSecretSource()
      return o
    }

    fun returnSourceViaFieldBad() {
      InferTaint.inferSensitiveSink(returnSourceViaField().f)
    }

    fun returnSourceViaParameter1(o: Obj) {
      o.f = InferTaint.inferSecretSource()
    }

    fun returnSourceViaParameter1Bad(o: Obj) {
      returnSourceViaParameter1(o)
      InferTaint.inferSensitiveSink(o.f)
    }

    fun returnSourceViaParameter2(o1: Obj, o2: Obj) {
      o2.f = o1.f
    }

    fun returnSourceViaParameter2Bad(o1: Obj, o2: Obj) {
      o1.f = InferTaint.inferSecretSource()
      returnSourceViaParameter2(o1, o2)
      InferTaint.inferSensitiveSink(o2.f)
    }

    fun returnSourceViaParameterOk(o1: Obj, o2: Obj) {
      o1.f = InferTaint.inferSecretSource()
      returnSourceViaParameter2(o2, o1)
      InferTaint.inferSensitiveSink(o2.f)
    }

    fun returnSourceViaGlobal() {
      global = InferTaint.inferSecretSource()
    }

    /** sink tests */
    fun callSinkParam1(param1: Any?, param2: Any?) {
      InferTaint.inferSensitiveSink(param1)
    }

    fun callSinkParam1Bad() {
      callSinkParam1(InferTaint.inferSecretSource(), null)
    }

    fun callSinkParam1Ok() {
      callSinkParam1(null, InferTaint.inferSecretSource())
    }

    fun callSinkParam2(param1: Any?, param2: Any?) {
      InferTaint.inferSensitiveSink(param2)
    }

    fun callSinkParam2Bad() {
      callSinkParam2(null, InferTaint.inferSecretSource())
    }

    fun callSinkParam2Ok() {
      callSinkParam2(InferTaint.inferSecretSource(), null)
    }

    fun callSinkOnFieldIndirect(param: Obj) {
      InferTaint.inferSensitiveSink(param.f)
    }

    fun callSinkOnFieldIndirectBad() {
      val obj = Obj()
      obj.f = InferTaint.inferSecretSource()
      callSinkOnFieldIndirect(obj)
    }

    fun callSinkOnGlobal() {
      InferTaint.inferSensitiveSink(global)
    }

    fun callSinkOnGlobalBad() {
      global = InferTaint.inferSecretSource()
      callSinkOnGlobal()
    }

    fun callSinkOnGlobalOk() {
      global = InferTaint.inferSecretSource()
      global = null
      callSinkOnGlobal()
    }

    fun setGlobalThenCallSinkBad() {
      global = InferTaint.inferSecretSource()
      callSinkOnGlobal()
    }

    val globalThenCallSink: Unit
      get() {
        val local = global
        InferTaint.inferSensitiveSink(global)
      }

    val globalThenCallSinkBad: Unit
      get() {
        global = InferTaint.inferSecretSource()
        globalThenCallSink
      }

    /** passthrough tests */
    fun singlePassthroughBad() {
      val source = InferTaint.inferSecretSource()
      val launderedSource = id(source)
      InferTaint.inferSensitiveSink(launderedSource)
    }

    fun doublePassthroughBad() {
      val source = InferTaint.inferSecretSource()
      val launderedSource1 = id(source)
      val launderedSource2 = id(launderedSource1)
      InferTaint.inferSensitiveSink(launderedSource2)
    }

    /** false positives: an ideal analysis would not report these, but we will */
    fun returnSourceConditional(b: Boolean): Any? {
      return if (b) InferTaint.inferSecretSource() else null
    }

    fun trackParamsOk() {
      InferTaint.inferSensitiveSink(returnSourceConditional(false))
    }

    fun reassignInCallee(o: Obj) {
      o.f = null
    }

    fun reassignInCallee() {
      val o = Obj()
      o.f = InferTaint.inferSecretSource()
      reassignInCallee(o)
      InferTaint.inferSensitiveSink(o.f)
    }

    fun relevantPassthrough(param: Any): Any {
      return param
    }

    fun irrelevantPassthrough(param: Any): Any {
      return param
    }

    // the following tests should show only "relevantPassthrough" in their traces
    fun irrelevantPassthroughsIntraprocedural(param: Any): Any {
      val irrelevant = irrelevantPassthrough(param)
      val source = InferTaint.inferSecretSource()
      val relevant = relevantPassthrough(source)
      InferTaint.inferSensitiveSink(relevant)
      return irrelevantPassthrough(relevant)
    }

    fun returnSourceIrrelevantPassthrough(param: Any): Any {
      val irrelevant = irrelevantPassthrough(param)
      val source = InferTaint.inferSecretSource()
      return relevantPassthrough(source)
    }

    fun irrelevantPassthroughsSourceInterprocedural(o: Any): Any {
      val irrelevant = irrelevantPassthrough(o)
      val source = returnSourceIrrelevantPassthrough(irrelevant)
      val relevant = relevantPassthrough(source)
      InferTaint.inferSensitiveSink(relevant)
      return irrelevantPassthrough(source)
    }

    fun callSinkIrrelevantPassthrough(param: Any): Any {
      val relevant = relevantPassthrough(param)
      InferTaint.inferSensitiveSink(relevant)
      return irrelevantPassthrough(param)
    }

    fun irrelevantPassthroughsSinkInterprocedural(o: Any?): Any {
      val source = InferTaint.inferSecretSource()
      val relevant = relevantPassthrough(source)
      callSinkIrrelevantPassthrough(relevant)
      return irrelevantPassthrough(relevant)
    }

    fun irrelevantPassthroughsSourceAndSinkInterprocedural(o: Any): Any {
      val irrelevant = irrelevantPassthrough(o)
      val source = returnSourceIrrelevantPassthrough(irrelevant)
      val relevant = relevantPassthrough(source)
      callSinkIrrelevantPassthrough(relevant)
      return irrelevantPassthrough(relevant)
    }

    fun callSinkVariadic(vararg params: Any?) {
      InferTaint.inferSensitiveSink(params)
    }

    fun callSinkVariadicBad() {
      callSinkVariadic(null, null, InferTaint.inferSecretSource())
    }

    fun callSinkThenDiverge(param: Any?) {
      InferTaint.inferSensitiveSink(param)
      while (true) {}
    }

    fun id2(o: Obj): Obj {
      return o
    }

    fun swapParams(o1: Any?, o2: Any?) {
      var o1 = o1
      o1 = o2
    }

    fun assignSourceToParam(o: Any?) {
      var o = o
      o = InferTaint.inferSecretSource()
    }

    fun swapParamsOk() {
      val notASource: Any? = null
      val source = InferTaint.inferSecretSource()
      swapParams(notASource, source)
      InferTaint.inferSensitiveSink(notASource)
    }

    fun assignSourceToParamOk() {
      val o: Any? = null
      assignSourceToParam(o)
      InferTaint.inferSensitiveSink(o)
    }
  }
}
